pkg/blockbuilder/blockbuilder.go

```diff
@@ -336,126 +331,134 @@ func (b *BlockBuilder) consumePartitions(ctx context.Context, part int32, mark t
 			}
 		})
 
-		if fetches.Empty() {
-			done = true
+		if fetches.Empty() && lag <= 0 {
+			level.Warn(b.logger).Log("msg", "got empty fetches from broker", "part", part)
+			break
 		}
 
-		fetches.EachPartition(func(ftp kgo.FetchTopicPartition) {
-			level.Info(b.logger).Log("msg", "consumed", "part", ftp.Partition, "hi", ftp.HighWatermark, "lo", ftp.LogStartOffset, "batch_size", len(ftp.Records))
+		recIter := fetches.RecordIter()
+		for !recIter.Done() {
+			rec := recIter.Next()
+
+			lag--
 
-			for _, rec := range ftp.Records {
+			// Stop consuming after we reached the cycleEnd marker.
+			// NOTE: the timestamp of the record is when the record was produced relative to distributor's time.
+			if rec.Timestamp.After(cycleEnd) {
+				if firstUncommittedRec == nil {
+					firstUncommittedRec = rec
+				}
+				done = true
+				break
+			}
+
+			// TODO(v): refactor block bounds check into a state machine:
+			// During catching up, if builder skipped samples, roll back the offset to first uncommited after compacting and starting next block
+			if rec.Timestamp.Before(blockStartAt) || rec.Timestamp.After(blockEndAt) {
 				// When BB is first deployed or if it is lagging behind, then it might consuming data from too much
 				// in the past. In which case if we try to consume all at once, it can overwhelm the system.
-				// So we break this into multiple block building cycles by resetting the block builder at intervals.
-				// TODO(codesome): the logic for this below is broken. Fix it. How can we determine the block ends when we are catching up?
-				if builder != nil && rec.Timestamp.After(resetBlockBuilderAt) {
-					if err := builder.compactAndClose(ctx, b.shipperDir()); err != nil {
-						level.Error(b.logger).Log("msg", "failed to compact and remove dbs", "part", part, "err", err)
-						// TODO(codesome): return err?
-						// TODO(codesome): add metric
+				// So we break this into multiple block building cycles.
+				if rec.Timestamp.After(blockEndAt) {
+					if err := builder.compact(ctx, b.shipperDir()); err != nil {
+						return fmt.Errorf("compact tsdb builder: %w", err)
 					}
-					builder = nil
 				}
-				if builder == nil {
-					builder = newTSDBBuilder(b.logger, b.limits, part, b.cfg.BlocksStorageConfig)
-					resetBlockBuilderAt = rec.Timestamp.Truncate(consumptionItvl).Add(consumptionItvl + timeBuffer)
+				blockStartAt, blockEndAt = blockBounds(rec.Timestamp, b.cfg.ConsumeInterval)
+			}
 
-					// TODO(codesome): verify. this can be wrong.
-					currEnd = rec.Timestamp.Truncate(consumptionItvl).UnixMilli()
-					lastEnd = currEnd - consumptionItvl.Milliseconds()
+			level.Debug(b.logger).Log("msg", "process record", "offset", rec.Offset, "rec", rec.Timestamp, "bmin", blockStartAt, "bmax", blockEndAt)
+
+			recProcessedBefore := false
```

Comment:
So that we don't forget it
```suggestion
			recProcessedBefore := false // TODO(codesome): get this from checkpoint
```